File: algo/strings/find_substring/rabin_karp/RabinKarpSearch.java
public class RabinKarpSearch {
    private static final int PRIME = 101;
    private static final int BASE = 256;
    
    public static int findSubstring(String text, String pattern) {
        int n = text.length();
        int m = pattern.length();
        int patternHash = 0;
        int textHash = 0;
        int h = 1;
        
        for (int i = 0; i < m - 1; i++) {
            h = (h * BASE) % PRIME;
        }
        
        for (int i = 0; i < m; i++) {
            patternHash = (BASE * patternHash + pattern.charAt(i)) % PRIME;
            textHash = (BASE * textHash + text.charAt(i)) % PRIME;
        }
        
        for (int i = 0; i <= n - m; i++) {
            if (patternHash == textHash) {
                int j = 0;
                while (j < m && text.charAt(i + j) == pattern.charAt(j)) {
                    j++;
                }
                if (j == m) {
                    return i;
                }
            }
            
            if (i < n - m) {
                textHash = (BASE * (textHash - text.charAt(i) * h) + text.charAt(i + m)) % PRIME;
                if (textHash < 0) {
                    textHash += PRIME;
                }
            }
        }
        return -1;
    }
    
    
    private static int calculateHash(String str, int start, int end) {
        int hash = 0;
        for (int i = start; i < end; i++) {
            hash = (BASE * hash + str.charAt(i)) % PRIME;
        }
        return hash;
    }
    
    public static void main(String[] args) {
        String[] testTexts = {
            "ABABDABACDABABCABCABCABCABC",
            "Hello World Hello",
            "AAAAAAA",
            "The quick brown fox jumps over the lazy dog"
        };
        
        String[] testPatterns = {
            "ABABCAB",
            "Hello",
            "AAA",
            "fox"
        };
        
        System.out.println("=== Rabin-Karp String Search Algorithm ===");
        System.out.println("Using BASE=" + BASE + " and PRIME=" + PRIME + "\n");
        
        for (int i = 0; i < testTexts.length; i++) {
            String text = testTexts[i];
            String pattern = testPatterns[i];
            
            System.out.println("Test " + (i + 1) + ":");
            System.out.println("Text: \"" + text + "\"");
            System.out.println("Pattern: \"" + pattern + "\"");
            
            int patternHash = calculateHash(pattern, 0, pattern.length());
            System.out.println("Pattern hash: " + patternHash);
            
            long startTime = System.nanoTime();
            int firstOccurrence = findSubstring(text, pattern);
            long endTime = System.nanoTime();
            
            if (firstOccurrence != -1) {
                System.out.println("First occurrence at index: " + firstOccurrence);
            } else {
                System.out.println("Pattern not found");
            }
            
            System.out.println("Time taken: " + (endTime - startTime) + " nanoseconds");
            System.out.println();
        }
    }
}

File: algo/strings/find_substring/rabin_karp/rabin_karp_search.py
def rabin_karp_search(text, pattern):
    """Rolling hash search: compare hash values"""
    if len(pattern) > len(text):
        return -1
    
    base = 256
    prime = 101
    
    # Compute hash values
    pattern_hash = 0
    text_hash = 0
    h = pow(base, len(pattern) - 1) % prime
    
    for i in range(len(pattern)):
        pattern_hash = (base * pattern_hash + ord(pattern[i])) % prime
        text_hash = (base * text_hash + ord(text[i])) % prime
    
    # Slide window and compare
    for i in range(len(text) - len(pattern) + 1):
        if pattern_hash == text_hash:
            if text[i:i + len(pattern)] == pattern:
                return i
        
        if i < len(text) - len(pattern):
            text_hash = (base * (text_hash - ord(text[i]) * h) + ord(text[i + len(pattern)])) % prime
            if text_hash < 0:
                text_hash += prime
    
    return -1

def test():
    text = "Programming is fun"
    pattern = "is"
    result = rabin_karp_search(text, pattern)
    print(f"Text: {text}")
    print(f"Pattern: {pattern}")
    print(f"Found at: {result}")

if __name__ == "__main__":
    test()